compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -337,13 +338,31 @@ trait TypeAssigner {
     }
   }
 
-  /** Substitute argument type `argType` for parameter `pref` in type `tp`,
-   *  skolemizing the argument type if it is not stable and `pref` occurs in `tp`.
+  /** Substitute argument type `argType` for parameter `pref` in type `tp`, but
+   *  take special measures if the argument is not realizable:
+   *  1. If the widened argument type is known to have good bounds,
+   *     substitute the skolemized argument type.
+   *  2. If the widened argument type is not known to have good bounds, eliminate all references
+   *     to the parameter in `tp`.
+   *  (2) is necessary since even with a skolemized type we might break subtyping if
+   *  bounds are bad. This could lead to errors not being detected. A test case is the second
+   *  failure in neg/i4031.scala
    */
-  def safeSubstParam(tp: Type, pref: ParamRef, argType: Type)(implicit ctx: Context) = {
+  def safeSubstParam(tp: Type, pref: ParamRef, argType: Type, initVariance: Int = 1)(implicit ctx: Context) = {
     val tp1 = tp.substParam(pref, argType)
-    if ((tp1 eq tp) || argType.isStable) tp1
-    else tp.substParam(pref, SkolemType(argType.widen))
+    if ((tp1 eq tp) || argType.isRealizable) tp1
+    else {
+      val widenedArgType = argType.widen
+      if (realizability(widenedArgType) == Realizable)
+        tp.substParam(pref, SkolemType(widenedArgType))
```

Comment:
`widenedArgType` can be realizable when `argType` isn't if argType is a lazy/erased value.
If `argType` is a normal value it's realizable always, even if `widenedArgType` isn't.

If neither is realizable, one could try harder, with an even wider type, but we don't have a reliable way to get at it — there are wider types (such as `Any`) but it's unclear there's a _smallest_ wider type, since the class hierarchy is a DAG.